adapters/repos/db/vector/hnsw/compress.go

```diff
@@ -46,84 +31,50 @@ func (h *hnsw) calculateOptimalSegments(dims int) int {
 }
 
 func (h *hnsw) Compress(cfg ent.PQConfig) error {
-	if h.isEmpty() {
-		return errors.New("Compress command cannot be executed before inserting some data. Please, insert your data first.")
-	}
-	err := h.initCompressedBucket()
-	if err != nil {
-		return fmt.Errorf("init compressed vector store: %w", err)
-	}
-
-	dims := int(h.dims)
+	h.compressActionLock.Lock()
+	defer h.compressActionLock.Unlock()
+	data := h.cache.All()
+	if cfg.Enabled {
+		if h.isEmpty() {
+			return errors.New("Compress command cannot be executed before inserting some data. Please, insert your data first.")
+		}
+		dims := int(h.dims)
 
-	if cfg.Segments <= 0 {
-		cfg.Segments = h.calculateOptimalSegments(dims)
-		h.pqConfig.Segments = cfg.Segments
-	}
+		if cfg.Segments <= 0 {
+			cfg.Segments = h.calculateOptimalSegments(dims)
+			h.pqConfig.Segments = cfg.Segments
+		}
 
-	h.pq, err = ssdhelpers.NewProductQuantizer(cfg, h.distancerProvider, dims)
-	if err != nil {
-		return fmt.Errorf("compress vectors: %w", err)
-	}
+		cleanData := make([][]float32, 0, len(data))
+		for _, point := range data {
+			if point == nil {
+				continue
+			}
+			cleanData = append(cleanData, point)
+		}
 
-	data := h.cache.All()
-	cleanData := make([][]float32, 0, len(data))
-	for _, point := range data {
-		if point == nil {
-			continue
+		var err error
+		h.compressor, err = compressionhelpers.NewPQCompressor(cfg, h.distancerProvider, dims, 1e12, h.logger, cleanData, h.store)
+		if err != nil {
+			return errors.Wrap(err, "Compressing vectors.")
+		}
+		h.commitLog.AddPQ(h.compressor.ExposeFields())
+	} else {
```

Comment:
I will add the check. Compress is only invoked when either PQ or BQ is enabled but this is a source for future bugs as is, as you have pointed out. So, it was working because the check was before but I totally agree we need to check this here too.